### ISSUE TITLE
Tone equalizer: Mouse wheel scrolling in picture doesn't change tone exposure after module activation

### ISSUE THREAD

**Issue by domosbg** [darktable-org/darktable#9834]
**Did you buy darktable from an application store ?**  -> No

**Describe the bug/issue**

**To Reproduce**
_Please provide detailed steps to reproduce the behaviour, for example:_
1. Picture is opened in darkroom
2. Click on the header of tone equalizer
Tone equalizer module is focused
![grafik](https://user-images.githubusercontent.com/88573684/129959032-c1b8a05a-7a6c-4e03-bddc-af63ee871b63.png)

In the area above the picture this help text is shown
![grafik](https://user-images.githubusercontent.com/88573684/129959139-f68756f5-4c79-4ddb-9dc2-0632aa95b85c.png)

But if you move the mouse wheel in the picture, picture is zoomed. Besides the mouse cursor doesn't change to the special cursor for tone exposure changing.

After you switches the module on by pressing the "power" button or you activate "display exposure mask" the special cursor is shown and the tone exposure changing by mouse wheel works:
![grafik](https://user-images.githubusercontent.com/88573684/129959939-b87e4d6c-b198-4fb5-8dad-01a3315dba30.png)

**Expected behavior**
If the module is focused and the help text for tone exposure change is shown, the mouse wheel operation in the picture should not zoom the picture but changes tone exposure.

**Platform**
_Please fill as much information as possible in the list given below. Please state "unknown" where you do not know the answer and remove any sections that are not applicable _

* darktable version : 3.7.0+684~g40421d57d
* OS                : Windows 10 // 19042.1165
* Memory : 16 GB
* Graphics card : Geforce RTX2070 + Intel UHD Graphics 630
* Graphics driver : 430.81
* OpenCL installed : yes
* OpenCL activated : yes

**Additional context**
_Please provide any additional information you think may be useful, for example:_

 - Can you reproduce with another darktable version(s)? **yes with version 3.6.0**
 - Are the steps above reproducible with a fresh edit (i.e. after discarding history)? **yes**

**Comment by Mark-64** [comment]
I am not sure this is a malfunction. Clicking on the module header focuses the module, displays the help text, but the cursor remains the normal one, which means the module is not active and therefore scrolling zooms the image.
Maybe the text could be more specific, adding something like "(activate the module first)"

**Comment by johnny-bit** [comment]
0.o I learn something new... module can be in focus but not enabled :D click on the "power" icon and it'll be enabled :)

Yeah... it needs "activate module" hint.

**Comment by domosbg** [comment]
Adding an "activate module first" hint would be the easiest way to fix it.

But actually the behaviour of tone equalizer differs from other modules like "crop" or "crop and rotate". 
In "crop" and "crop and rotate" after focusing the module the resize rectangle is shown. There is no need to activate the module first.
On the other hand in "crop" and "crop and rotate" you can always zoom because there is no mouse wheel action on the picture from the module. So if the priorization is on zooming then the actual behaviour is ok. If the priorization is on the modules function then the "tone exposure cursor" should be activated if module is focused. Alternatively you could change tone exposure change to "Shift + Mouse Wheel" or something else. But that makes it much more complicate as it actualy is. I think there is no perfect solution that fits all needs.

Maybe my workflow is different from the normal one, but I never use the "power" icon to initial switch on a module. I click on the header, make my modification and the module is switched on automatically. I only use the "power" icon if I want to activate/deactivate a module to check the effect on my picture. Because of that for me the handling of the "tone equalizer" differs from all the other modules I use.

**Comment by elstoc** [comment]
Modifying any module parameter will enable it. In the case of most modules that's changing a slider/combobox, but those modules that can have parameters set by on-image interfaces work differently. I always crop by focusing the module first and then it's only enabled when I actually change something. It feels like Tone Equalizer should probably work the same way.

**Comment by johnny-bit** [comment]
Now taht no-activity bot bumped issue I think #9565 and this one are connected conceptually... In #9565 the issue is that module has focus when user expects it not to have and the inverse is true here...

**Comment by Mark-64** [comment]
I am looking into this issue and I believe I am onto something, but I need help.
Comparing Tone Equalizer with Graduated Density, which seems to have the desired behavior of showing the overlay cursor on focus and activate the module when something is changed, I see the following:
1. in both modules, overlay is drawn in `gui_post_expose()` callback

2.  In `gui_post_expose()`, Tone Equalizer has this check at the beginning
```

  dt_iop_gui_enter_critical_section(self);
  const int fail = (!g->cursor_valid || !g->interpolation_valid || !g->luminance_valid || dev->pipe->processing || !sanity_check(self) || !g->has_focus);
  dt_iop_gui_leave_critical_section(self);
  if(fail) return;
```
Now, g->cursor_valid is only set to TRUE in `mouse_moved()` callback, if the cursor is inside the central picture, this in order to draw the overlay only when appropriate

3. The problem is that, while in Graduated Density the `mouse_moved()` callback is called as soon the module is in focus, for Tone Equalizer this doesn't happen, the module need to be activated.

The question for someone more expert than me is: where do change this behavior?

**Comment by Mark-64** [comment]
No, wait.
In Tone Equalizer, mouse_moved():
```
if(!self->enabled) return 0;
```
so it is done on purpose. The question is now, why? I will investigate further

**Comment by TurboGit** [comment]
What we usually do is to enable the module as soon some settings are changed. I think we need to do that here

**Comment by TurboGit** [comment]
What we usually do is to enable the module as soon some settings are changed. I think we need to do that here

**Comment by Mark-64** [comment]
In Tone Equalizer

```
int mouse_moved(struct dt_iop_module_t *self, double x, double y, double pressure, int which)
{
  dt_develop_t *dev = self->dev;
  dt_iop_toneequalizer_gui_data_t *g = (dt_iop_toneequalizer_gui_data_t *)self->gui_data;
  if(!self->enabled) return 0;

  dt_iop_gui_enter_critical_section(self);
  const int fail = (!sanity_check(self) || !g->luminance_valid);
  dt_iop_gui_leave_critical_section(self);
  if(fail) return 0;

```
so not only the module must be enabed, but also `g->luminance_valid` must be `TRUE`.
With my little knowledge of the module, there seems to be a luminance mask which is pre-computed and stored in a buffer.
Without that, the `mouse_moved()` callback returns and the special cursor is not displayed.
I am afraid I can't go further, I don't even know if fixing that is possible or desirable.

**Comment by johnny-bit** [comment]
Ah, module must be enabled because enabling module calculates the default mask (thus making luminance_valid)

IF we'd want to have the cursor working it should then show not necessarly `NaN EV` but maybe nothing or `?` and then on scroll force enable module?

**Comment by Mark-64** [comment]
Maybe, let me try

**Comment by Mark-64** [comment]
@domosbg can you test #9845 ?

**Comment by domosbg** [comment]
Thanks a lot for your work. I would really like to test it, but TBH actually I have no idea how to do it. I think a have to compile darktable from the sources with your changes?

Is this a good start to learn how to do it: https://github.com/darktable-org/darktable#easy-way

I will try to compile it. If it works I will test your changes.

**Comment by domosbg** [comment]
Easier than expected. I managed it to compile darktable.

Tested your changes in tone equalizer. It works as expected by me. And I could not identify any problems.

**Comment by Mark-64** [comment]
@domosbg you are recruited in the DT dev community as a tester then ;-)

**Comment by domosbg** [comment]
I will try my best ;-)

Should I close the issue, because the problem is fixed? Or leave it open and it is closed automatically if 3.8 is released?

**Comment by johnny-bit** [comment]
> Or leave it open and it is closed automatically if 3.8 is released?

Leave it open. It'll be automagically closed once #9845 passes review process and is merged.
